### ISSUE TITLE
RTQ gets blocked by bad consumers.

### ISSUE THREAD

**Issue by lordnull** [basho/riak_repl#443]
TLDR: A consumer can block the rtq from processing further messages indefinitely.

A client site encountered a situation where riak-repl status was hanging. This was tracked down to the rtq not responding to status (blocking the caller infinitely). This was further tracked down to the rtq being in the middle of handling a pull made by an rtsource_helper process. A helper process has a pull fulfilled by having the rtq make a call into the helper. The helper in this case was blocked by sending a heartbeat; i.e.: prim_inet:send/3 was blocking indefinitely it seems.

In this short term this was resolved by killing the helper process. However, the helper process should not have been able to block the rtq completely.

Reason for the blocking pull from rtq to helper is to ensure delivery to that consumer and move it from 'pending' to 'waiting for ack'. Simply changing the deliver fun (used in the pull) from a call to a cast removes that guarantee. Changing the timeout from infinite to something not infinite means it have have been successfully delivered, yet reported as a failure.

**Comment by bookshelfdave** [comment]
Fixed in https://github.com/basho/riak_repl/pull/461
